rfcs/063-catalogue_pipeline_lambdas/README.md

```diff
@@ -0,0 +1,139 @@
+
+  Service deployment at present relies on the ECS deployment APIs which take some time to determine that a service is "stable", there is no such process for AWS Lambda and deployments are almost instant.
+
+- **Align with concepts pipeline deployment:**  
+
+   The concepts-pipeline was designed and built around lambdas. It packages 2 different versions of the aggregator and recorder services, one for each use case for the pipeline: bulk, for complete concepts reindex and SQS, which handles low-volume 
+   works updates published to `<live_catalogue_pipeline>_ingestor_works_output` SNS topic. The ingestor service only runs when needed, eg. when we want to refresh the `authoritative-concepts` index.  
+
+   This allows for most efficient use of the infrastructure/resources and avoids one of the oft-cited complexity/downside in the catalogue-pipeline, namely that the application tries to handle two quite different use cases.  
+
+   The different versions of the aggregator and recorder are also configured for their purpose, thus making the infrastructure code easier to read and understand.
+
+## Suggested solution
+
+We could incrementally refactor the catalogue-pipeline services to use lambdas instead of ECS. Some considerations are:
+
+ - We currently build container images to packages our code for deployment, we may wish to continue doing so for simplicity. See the documentation around [packaging and deploying container based Java Lambdas](https://docs.aws.amazon.com/lambda/latest/dg/java-image.html) that we would need to use to achieve this.
+
+- Lambda has built-in scaling capabilities which will allow us to simplify the infrastructure by removing scaling configurations and Cloudwatch alarms on the queues. We would still use the `reindexing_state` to adjust data store provisioning. We 
+  could also leverage SQS lambda features to protect the data stores (eg. elasticsearch index at the ingestor stage) from overloading by setting a maximum concurrency on the event source.
+   
+- We can leverage SQS event source and lambda features to batch up to 10000 messages before invoking the function (useful in the context of the batcher service). This will make it possible to remove the batching mechanism that exists inside the 
+  services, making it easier to run them locally using a [lambda runtime emulator](https://github.com/aws/aws-lambda-runtime-interface-emulator).
+
+  Error handling could use [Partial batch responses](https://docs.aws.amazon.com/lambda/latest/dg/services-sqs-errorhandling.html#services-sqs-batchfailurereporting).
+  
+- Lambda is generally considered to be cost-effective for workloads with intermittent or unpredictable traffic patterns (1ms for 1024MB: $0.0000000167).
+
+- Processing speed could be slowed down by [AWS Lambda cold starts](https://aws.amazon.com/blogs/compute/operating-lambda-performance-optimization-part-1/) but this can be optimised by provisioning concurrency.
```

Comment:
It's good to mention this as a potential mitigation, but it may be unnecessary and this may not be the best place to start looking in order to reduce overall latency.
 
During normal running, this model is likely to be at least as fast as the current approach (because an ECS is also quite likely to have spun down and now needs to come up again), and real time processing is not expected anyway. 

When it's busy, the Lambdas are likely to be reused.

We may get some good results from looking at what we can hold onto in memory between runs.  Similarly, we may be able to do away with Akka in some places, which would reduce both the package size and the internal startup time.

Reply:
Yes, good point that there are other optimisations besides provisioned concurrency. How to hold onto things in-memory for containerised lambdas is not clear to me, and is further mudied by these being JVM based applications ... I guess it may be as easy as putting variables in global scope?

Besides the options around [static initialization](https://docs.aws.amazon.com/lambda/latest/operatorguide/static-initialization.html#:~:text=The%20INIT%20code%20runs%20when,uses%20a%20warm%20execution%20environment.) we could also use /tmp?

I'll add this to the doc.